pkg/util/spill/agg_spill_util.go

```diff
@@ -0,0 +1,349 @@
+// However, `Enum` type is always deserialized with fix length types.
+// So, there is no need to return deserialized bytes so far.
+func DeserializeEnum(buf []byte, pos int64) types.Enum {
+	retValue := types.Enum{}
+	retValue.Value = DeserializeUint64(buf, pos)
+	retValue.Name = string(buf[pos+uint64Len:])
+	return retValue
+}
+
+// SerializeBool serializes bool type
+func SerializeBool(value bool, tmpBuf []byte) {
+	*(*bool)(unsafe.Pointer(&tmpBuf[0])) = value
+}
+
+// SerializeInt serializes int type
+func SerializeInt(value int, tmpBuf []byte) {
+	*(*int)(unsafe.Pointer(&tmpBuf[0])) = value
+}
+
+// SerializeInt8 serializes int8 type
+func SerializeInt8(value int8, tmpBuf []byte) {
+	*(*int8)(unsafe.Pointer(&tmpBuf[0])) = value
+}
+
+// SerializeUint8 serializes uint8 type
+func SerializeUint8(value uint8, tmpBuf []byte) {
+	*(*uint8)(unsafe.Pointer(&tmpBuf[0])) = value
+}
+
+// SerializeInt32 serializes int32 type
+func SerializeInt32(value int32, tmpBuf []byte) {
+	*(*int32)(unsafe.Pointer(&tmpBuf[0])) = value
+}
+
+// SerializeUint32 serializes uint32 type
+func SerializeUint32(value uint32, tmpBuf []byte) {
+	*(*uint32)(unsafe.Pointer(&tmpBuf[0])) = value
+}
+
+// SerializeUint64 serializes uint64 type
+func SerializeUint64(value uint64, tmpBuf []byte) {
+	*(*uint64)(unsafe.Pointer(&tmpBuf[0])) = value
+}
+
+// SerializeInt64 serializes int64 type
+func SerializeInt64(value int64, tmpBuf []byte) {
+	*(*int64)(unsafe.Pointer(&tmpBuf[0])) = value
+}
+
+// SerializeFloat32 serializes float32 type
+func SerializeFloat32(value float32, tmpBuf []byte) {
+	*(*float32)(unsafe.Pointer(&tmpBuf[0])) = value
+}
+
+// SerializeFloat64 serializes float64 type
+func SerializeFloat64(value float64, tmpBuf []byte) {
+	*(*float64)(unsafe.Pointer(&tmpBuf[0])) = value
+}
+
+// SerializeInterface serialize interface type and return the number of bytes serialized
+func SerializeInterface(value interface{}, varBuf *[]byte, tmpBuf []byte) {
+	switch v := value.(type) {
+	case bool:
+		*varBuf = append(*varBuf, BoolType)
+		if v {
+			*varBuf = append(*varBuf, byte(1))
+		} else {
+			*varBuf = append(*varBuf, byte(0))
+		}
+	case int64:
+		*varBuf = append(*varBuf, Int64Type)
+		SerializeInt64(v, tmpBuf)
+		*varBuf = append(*varBuf, tmpBuf[:int64Len]...)
+	case uint64:
+		*varBuf = append(*varBuf, Uint64Type)
+		SerializeUint64(v, tmpBuf)
+		*varBuf = append(*varBuf, tmpBuf[:uint64Len]...)
+	case float64:
+		*varBuf = append(*varBuf, FloatType)
+		SerializeFloat64(v, tmpBuf)
+		*varBuf = append(*varBuf, tmpBuf[:float64Len]...)
+	case string:
+		*varBuf = append(*varBuf, StringType)
+		vLen := int64(len(v))
+		SerializeInt64(vLen, tmpBuf)
+		*varBuf = append(*varBuf, tmpBuf[:int64Len]...)
+		*varBuf = append(*varBuf, v...)
+	case types.BinaryJSON:
+		*varBuf = append(*varBuf, BinaryJSONType)
+		varBufLenBeforeSerializeJSON := int64(len(*varBuf))
+
+		// Add padding for seialization
+		*varBuf = append(*varBuf, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1)
```

Comment:
> why do we need this?

Because `SerializeXXX` supposes the passed buffer has enough space to put serialized bytes, and we should ensure this guarantee.